src/Microsoft.AspNetCore.Hosting/Internal/HostingApplicationData.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Microsoft.Extensions.FileProviders;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    public class HostingApplicationData : IHostingApplicationData
+    {
+        public string ApplicationDataPath { get; set; }
+
+        public IFileProvider ApplicationDataFileProvider { get; set; }
+
+        internal void Initialize(WebHostOptions options)
+        {
+            CreateHostingApplicationData(options.ApplicationDataPath);
+        }
+
+        private void CreateHostingApplicationData(string basePath)
+        {
+            var appDataPath = ResolveApplicationDataPath(basePath);
+            if (appDataPath != null)
+            {
+                ApplicationDataPath = appDataPath;
+                ApplicationDataFileProvider = new PhysicalFileProvider(appDataPath);
+            }
+        }
+
+        private string ResolveApplicationDataPath(string applicationDataPath)
+        {
+            var directoryInfo = GetOrCreateDirectory(applicationDataPath);
+            if (directoryInfo != null)
+            {
+                return directoryInfo.FullName;
+            }
+
+            // Environment.GetFolderPath returns null if the user profile isn't loaded.
+            var localAppDataFromSystemPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            var localAppDataFromEnvPath = Environment.GetEnvironmentVariable("LOCALAPPDATA");
+            var userProfilePath = Environment.GetEnvironmentVariable("USERPROFILE");
+            var homePath = Environment.GetEnvironmentVariable("HOME");
+
+            // To preserve backwards-compatibility with 1.x, Environment.SpecialFolder.LocalApplicationData
+            // cannot take precedence over $LOCALAPPDATA and $HOME/.aspnet on non-Windows platforms.
+            // We do this in here too to keep consistency with the locations where data protection stores keys.
+            directoryInfo = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ?
+                GetOrCreateDirectory(GetFolderPath(localAppDataFromSystemPath, "ASP.NET")) :
+                null;
+
+            directoryInfo = directoryInfo ??
+                GetOrCreateDirectory(GetFolderPath(localAppDataFromEnvPath, "ASP.NET")) ??
+                GetOrCreateDirectory(GetFolderPath(userProfilePath, "AppData", "Local", "ASP.NET")) ??
```

Comment:
Some of these literals are used multiple times ("ASP.NET" folder). Aren't there constants for these well-known paths/folders ?